src/inmanta/ast/statements/__init__.py

```diff
@@ -149,54 +193,258 @@ def get_dataflow_node(self, graph: DataflowGraph) -> dataflow.NodeReference:
+    :ivar statement: The assignment statement that led to this promise.
+    """
+
+    instance: "Reference"
+    attribute: str
+    statement: "SetAttribute"
+
+    def get_root_variable(self) -> str:
+        """
+        Returns the name of the variable at the start of the attribute traversal chain. e.g. for a.b.c.d, returns "a". Includes
+        namespace information if specified in the original reference.
+        """
+        return self.instance.get_root_variable().name
+
+    def schedule(self, responsible: DynamicStatement, resolver: Resolver, queue: QueueScheduler) -> "EagerPromise":
+        """
+        Schedule the acquisition of this promise in a given dynamic context: set up a waiter to wait for the referenced
+        ResultVariable to exist, then acquire the promise.
+
+        :param responsible: The statement responsible for this eager promise, i.e. the statement that will fulfill it once it
+            will make no further progression.
+        """
+        dynamic: "EagerPromise" = EagerPromise(self, responsible)
+        hook: VariableReferenceHook = VariableReferenceHook(
+            self.instance,
+            self.attribute,
+            variable_resumer=dynamic,
+        )
+        self.statement.copy_location(hook)
+        waiter: Waiter = hook.schedule(resolver, queue)
+        dynamic.set_waiter(waiter)
+        return dynamic
+
+
+class EagerPromise(VariableResumer):
+    """
+    Dynamic node for eager promising (stateful). Eagerly acquires a progression promise on a variable when it becomes available.
+    Fulfilling this promise aborts the waiter if it has not finished yet, otherwise it fulfills the acquired progression
+    promise.
+    """
+
+    def __init__(self, static: StaticEagerPromise, responsible: DynamicStatement) -> None:
+        super().__init__()
+        self.static: StaticEagerPromise = static
+        self.responsible: DynamicStatement = responsible
+        self._waiter: Optional[Waiter] = None
+        self._promise: Optional[ProgressionPromise] = None
+
+    def set_waiter(self, waiter: Waiter) -> None:
+        self._waiter = waiter
+
+    def _acquire(self, variable: ResultVariable) -> None:
+        """
+        Entry point for the ResultVariable waiter: actually acquire the promise
+        """
+        if self._waiter is None:
+            # already fulfilled, no need to acquire progression promise anymore
+            return
+        assert self._promise is None
+        self._promise = variable.get_progression_promise(self.responsible)
+        self._waiter = None
+
+    def fulfill(self) -> None:
+        """
+        If a promise was already acquired, fulfills it, otherwise cancels the waiter so no new promise is acquired when the
+        variable becomes available.
+        """
+        if self._waiter is not None:
+            # waiter is still waiting, remove it
+            self._waiter.queue.remove_from_all(self._waiter)
```

Comment:
I wonder if this is a good idea, performance wise. 

it also doesn't 'cancel' the waiter in any way. 
(remove_from_all is iirc pure bookkeeping)

Reply:
Didn't seem to have a noticable performance impact but I removed it anyway because it makes it simpler to understand.